### PR TITLE
RK-12523 Bitbucket support fetch by file path

### PR DESCRIPTION
**Pull request by hopemanryan**
### Type of change

- [ ] Improvement
- [ ] Bugfix
- [ ] Refactor
- [x] New feature

### Size of change

- [ ] Small (<3 files)
- [ ] Medium (new file(s) or >3 files)
- [x] Large (>10 files or major change)

### Description of changes
Added support for fetching per path for bitbucket
 Fetching by path requires the same details like fetching the entire tree but with an additional key "filePath"
which returns only the files/folders under that strictly under that path.

### QA process
1. Set up on-prem bitbucket repo and add any repo to it.
2. Send a request to the explorerook under fileTreeByPath
3. see that the only files received are under the requested path.
4. check that original query fileTree  works like before

### CODE REVIEW

File: src/BitBucketOnPrem.ts [hunk 1]
```diff
@@ -25,57 +36,114 @@ export interface BitBucketOnPremInput {
...
+            }).toString();
+
+        const sluggedUrl = addSlugToUrl(fileTreeUrl, filePath);
 
+        logger.debug("Getting files for", {projectKey, repoName, url, commit});
+        let isLastPage = false;
+        let start = 0;
+        const files: string[] = [];
+        while (!isLastPage) {
+            const res = await fetchNoCache(`${sluggedUrl}?start=${start}&limit=${FETCH_LIMIT}`, {
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:76]: Should you not lower the page size?

File: src/BitBucketOnPrem.ts [hunk 2]
```diff
@@ -25,57 +36,114 @@ export interface BitBucketOnPremInput {
...
+            });
+            const fileTreeResponse: any = await res.json();
+
+            const {children} = fileTreeResponse || {};
+            const {values} = children || {};
+            if (Array.isArray(values)) {
+                for (const item of values) {
+                    const {type, path} = item;
+                    const {name} = path || {};
+                    files.push(`${filePath && false ? filePath + "/" : ""}${name}${type === FILE_TYPE.DIRECTORY ? "/" : ""}`);
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:89]: && false?

**Review by Urook — CHANGES_REQUESTED** [review]

**Review by Urook — COMMENTED** [review]
You must raise the patch version in package.json

**Review by Urook — CHANGES_REQUESTED** [review]
Is there no way to avoid using the permenant is large storage? I would rather it be on the spot

**Comment by hopemanryan** [comment]
Well if we want to keep the setting even if the user closes the app than it's that or saving it to a local file ( which is to much of an over head I think )

File: src/BitBucketOnPrem.ts (hunk 1 above)
Comment (hopemanryan) [reply src/BitBucketOnPrem.ts:76]: 1. the high limit is an actual performance benefit it seems in bitbucket (@DavidRBiderman  had the pleasure of figuring that one out)  
2. IMO the client should decide if its large or not and ask explorerook to set it as large, but we can discuss this one

File: src/BitBucketOnPrem.ts [hunk 3]
```diff
@@ -213,3 +281,31 @@ export const getCommitDetailsFromBitbucket = async ({url, accessToken, projectKe
...
+const addSlugToUrl = (url: string, slug: string): string => {
+    if (!slug) {
+        return url;
+    }
+    return `${url}/${slug}`;
+};
+
+
+export const saveRepoAsLarge = (repoName: string, url: string): boolean => {
+    const fromLocalStorage: string = localStorage.getItem(LOCAL_KEY);
```
[the remark this answers is not in the document]
Comment (hopemanryan) [reply src/BitBucketOnPrem.ts:294]: I dont understand this comment, can you please explain  ?

File: graphql/schema.graphql [hunk 4]
```diff
@@ -123,13 +123,15 @@ enum OnPremSourceProvider {
...
   fileTree(args: BitbucketInput!): [String]
+  fileTreeByPath(args: BitbucketInput!): [String]
   user(args: BitbucketInput!): BitbucketUser
   projects(args: BitbucketInput!): [BitbucketProject]
   repos(args: BitbucketInput!): [BitbucketRepo]
   commits(args: BitbucketInput!): [BitbucketCommit]
   commit(args: BitbucketInput!): BitbucketCommit
   branches(args: BitbucketInput!): [BitbucketBranch]
   file(args: BitbucketInput!): String
+
```
Comment (Urook) [inline graphql/schema.graphql:134]: Delete newline

File: src/BitBucketOnPrem.ts [hunk 5]
```diff
@@ -25,57 +34,105 @@ export interface BitBucketOnPremInput {
     args: BitbucketOnPrem;
 }
 
+
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:37]: Newline

File: src/BitBucketOnPrem.ts [hunk 6]
```diff
@@ -25,57 +34,105 @@ export interface BitBucketOnPremInput {
...
         requestInit.credentials = 'omit';
     }
-    return fetch(requestInfo.toString().replace('scm/', '').replace('scm%2F',''), requestInit);
-}
+    return fetch(requestInfo.toString().replace('scm/', '').replace('scm%2F', ''), requestInit);
+};
+
+export const getFileTreeByPath =
+    async ({url, accessToken, projectKey, repoName, commit, filePath}: BitbucketOnPrem): Promise<string[]> => {
+        const templateUrl: TREE_FETCH_URL = TREE_FETCH_URL.BY_PATH;
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:53]: Why do you need this?

File: src/BitBucketOnPrem.ts [hunk 7]
```diff
@@ -25,57 +34,105 @@ export interface BitBucketOnPremInput {
...
+        // build url without filePath (Url slug needs to be "src/folder" but UrlAssembler will turn it into unicode)
+        const fileTreeUrl = UrlAssembler(url).template(templateUrl)
+            .param({
+                projectKey,
+                repoName,
+            }).toString();
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:59]: You are missing a query parameter for the commit

File: src/api.ts [hunk 8]
```diff
@@ -7,7 +7,7 @@ import {
   getBranchesForRepoFromBitbucket,
   getCommitDetailsFromBitbucket,
   getCommitsForRepoFromBitbucket,
-  getFileContentFromBitbucket,
+  getFileContentFromBitbucket, getFileTreeByPath,
```
Comment (Urook) [inline src/api.ts:10]: 2 lines

File: src/BitBucketOnPrem.ts [hunk 9]
```diff
@@ -25,57 +34,105 @@ export interface BitBucketOnPremInput {
...
 
+        const sluggedUrl = addSlugToUrl(fileTreeUrl, filePath);
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:61]: Do this before the assembler, as you will need the "at" query param

**Review by Urook — CHANGES_REQUESTED** [review]

File: src/BitBucketOnPrem.ts [hunk 10]
```diff
@@ -30,52 +39,102 @@ export interface BitBucketOnPremInput {
...
         requestInit.credentials = 'omit';
     }
-    return fetch(requestInfo.toString().replace('scm/', '').replace('scm%2F',''), requestInit);
-}
+    return fetch(requestInfo.toString().replace('scm/', '').replace('scm%2F', ''), requestInit);
+};
+
+export const getFileTreeByPath =
+    async ({url, accessToken, projectKey, repoName, commit, filePath}: BitbucketOnPrem): Promise<string[]> => {
+        const templateUrl: string =  addSlugToUrl(TREE_FETCH_URL.BY_PATH, filePath);
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:52]: double space before addSlugToUrl.
Also, why is the TREE_FETCH_URL const needed?

**Review by Urook — CHANGES_REQUESTED** [review]

File: src/BitBucketOnPrem.ts [hunk 11]
```diff
@@ -30,52 +35,102 @@ export interface BitBucketOnPremInput {
...
     }
-    return fetch(requestInfo.toString().replace('scm/', '').replace('scm%2F',''), requestInit);
-}
+    return fetch(requestInfo.toString().replace('scm/', '').replace('scm%2F', ''), requestInit);
+};
+
+export const getFileTreeByPath =
+    async ({url, accessToken, projectKey, repoName, commit, filePath}: BitbucketOnPrem): Promise<string[]> => {
+        const templateUrl: string = addSlugToUrl('rest/api/1.0/projects/:projectKey/repos/:repoName/browse', filePath);
+        // build url without filePath (Url slug needs to be "src/folder" but UrlAssembler will turn it into unicode)
```
Comment (Urook) [inline src/BitBucketOnPrem.ts:49]: Wait, did not see this comment. Can you check if it works? It still needs to happen before UrlAssembler imo but maybe there is a way to tell it not to parse it to unicode.

**Review by Urook — CHANGES_REQUESTED** [review]

File: src/BitBucketOnPrem.ts (hunk 11 above)
Comment (hopemanryan) [reply src/BitBucketOnPrem.ts]: comment is no longer needed since we add the slug to the template before going through the url assembler

**Review by Urook — APPROVED** [review]

**Pull request merged**